examples/sod/ArborX_DetailsSOD.hpp

```diff
@@ -0,0 +1,388 @@
+}
+
+// Compute critical bins
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<int *, MemorySpace> computeSODCriticalBins(
+    ExecutionSpace const &exec_space, float rho, float rho_ratio,
+    Kokkos::View<double **, MemorySpace> sod_halo_bin_masses,
+    Kokkos::View<int **, MemorySpace> sod_halo_bin_counts,
+    Kokkos::View<float **, MemorySpace> sod_halo_bin_outer_radii)
+{
+  Kokkos::Profiling::pushRegion("ArborX::SOD::compute_critical_bins");
+
+  auto const num_halos = sod_halo_bin_masses.extent(0);
+  int const num_sod_bins = sod_halo_bin_masses.extent(1);
+
+  Kokkos::View<int *, MemorySpace> critical_bin_ids(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing,
+                         "ArborX::SOD::critical_bin_ids"),
+      num_halos);
+  Kokkos::deep_copy(critical_bin_ids, -1);
+  Kokkos::parallel_for(
+      "ArborX::SOD::compute_critical_bins",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_halos),
+      KOKKOS_LAMBDA(int halo_index) {
+        int &critical_bin_id = critical_bin_ids(halo_index);
+
+        double accumulated_mass = 0.;
+        for (int bin_id = 0; bin_id < num_sod_bins; ++bin_id)
+        {
+          accumulated_mass += sod_halo_bin_masses(halo_index, bin_id);
+          float outer_radius = sod_halo_bin_outer_radii(halo_index, bin_id);
+
+          float bin_rho_ratio_int = 0;
+          if (outer_radius > 0)
+          {
+            auto volume = 4.f / 3 * M_PI * pow(outer_radius, 3);
+            bin_rho_ratio_int = (accumulated_mass / volume) / rho;
+          }
+
+          if (bin_rho_ratio_int <= rho_ratio)
+          {
+            critical_bin_id = bin_id;
+            break;
+          }
+        }
+
+        if (critical_bin_id < 0)
+        {
+          printf("%d (halo tag ?): max radius is not big enough, will "
+                 "underestimate\n",
+                 halo_index);
+          critical_bin_id = num_sod_bins - 1;
+          while (sod_halo_bin_counts(halo_index, critical_bin_id) == 0 &&
+                 critical_bin_id > 0)
+            --critical_bin_id;
+        }
+        else if (critical_bin_id == 0)
+        {
+          printf("%d (halo tag ?): min radius is not small enough, will "
+                 "overestimate\n",
+                 halo_index);
+          while (sod_halo_bin_counts(halo_index, critical_bin_id) == 0 &&
+                 critical_bin_id < num_sod_bins - 1)
+            ++critical_bin_id;
+        }
+
+        // Capture the cases where normal critical bin is empty, and that
+        // extra radius meets the rho_ratio criteria
+        while (sod_halo_bin_counts(halo_index, critical_bin_id) == 0)
+          --critical_bin_id;
+
+        // This should not happen but protecting against no zero bins for all up
+        // to the critical bin
+        if (critical_bin_id < 0)
+          printf(
+              "%d (halo tag ?): error, not even the first bin has particles\n",
+              halo_index);
+      });
```

Comment:
The whole thing is really really convoluted. Mainly because there may be bins with 0 particles in them, in which case CosmoTools tries to find one with particles. You would think Cosmotools had those unit tests, but I don't think they do.

Overally, my hope is to convince HACC developers to switch to the second version that is a lot more streamlined and comes with some guarantees. We'll see if that's successful.